independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanProcessor.java

```diff
@@ -509,6 +512,34 @@ public Predicate<DotName> getInjectionPointAnnotationsPredicate() {
         return injectionPointAnnotationsPredicate;
     }
 
+    private Set<DotName> findSingleContextNormalScopes() {
+        Set<DotName> ret = new HashSet<>();
+        Set<DotName> builtinScopes = Set.of(BuiltinScope.REQUEST.getName());
+        Set<DotName> customScopes = beanDeployment.getCustomContexts().keySet().stream().filter(ScopeInfo::isNormal)
+                .map(ScopeInfo::getDotName)
+                .collect(Collectors.toSet());
+        Set<DotName> alterableScopes = customAlterableContexts.getRegistered().stream()
+                .filter(cac -> Boolean.TRUE.equals(cac.isNormal)).map(cac -> cac.scopeAnnotation)
+                .map(DotName::createSimple)
+                .collect(Collectors.toSet());
+        for (DotName s : builtinScopes) {
+            if (!customScopes.contains(s) && !alterableScopes.contains(s)) {
+                ret.add(s);
+            }
+        }
+        for (DotName s : customScopes) {
+            if (!builtinScopes.contains(s) && !alterableScopes.contains(s)) {
+                ret.add(s);
+            }
+        }
+        for (DotName s : alterableScopes) {
+            if (!builtinScopes.contains(s) && !customScopes.contains(s)) {
+                ret.add(s);
+            }
+        }
```

Comment:
This seems fairly useless. I believe this

```java
ret.addAll(builtinScopes);
ret.addAll(customScopes);
ret.addAll(alterableScopes);
```

would do the same thing, even performance-wise (because the checks we do need to perform the hash lookups as well), it would just be easier to understand.

Reply:
I don't think so. We need to find the scopes for which a single context exists. There are three sets of contexts and we need to find a scope from each set that is not contained in other two sets. It's not exatly symmetric difference, but I don't know if there's a name for this set operation...

In any case, your code would not work. Image we have `{AppScoped}`, `{AppScoped, RequestScoped}` and `{}`. Your code would return `{AppScoped, RequestScoped}` which is not correct, because there are two context implementations for `AppScoped`. The correct result is `{RequestScoped}`.

Maybe I'm missing something?

Reply:
Oh you are absolutely right! But you're making the same mistake actually :-) The `customScopes` and `alterableScopes` are `Set`s, so you will not detect when there are multiple contexts for a single scope among custom contexts (same for custom alterable contexts).

Reply:
I see that `BeanDeployment` uses a `Map` to store custom contexts, so if someone adds multiple `ContextRegistrar`s for the same scope, only one will survive, so we don't generally support registering multiple contexts for the same scope through that mechanism and it's safe to use a `Set` here. I don't think that applies to `CustomAlterableContexts`.

Reply:
Yes, I mentioned this already (even on the zulip discussion) - ArC does not support multiple custom contexts for the same scope. As for the `CustomAlterableContexts` - yes, in theory you could register multiple contexts for the same scope but given that our integration is far from being perfect (for example `CustomScopeAnnotationsBuildItem` and `io.quarkus.arc.processor.BuildExtension.Key.SCOPES` would not reflect those contexts) I think that our approximation is good enough for now...

Reply:
Well, this is an optimization, so either we need to precise, or the approximation needs to be conservative. WDYT about this implementation?

```java
private Set<DotName> findSingleContextNormalScopes() {
    Map<DotName, Integer> contextsForScope = new HashMap<>();
    // built-in contexts
    contextsForScope.put(BuiltinScope.REQUEST.getName(), 1);
    // custom contexts
    beanDeployment.getCustomContexts()
            .keySet()
            .stream()
            .filter(ScopeInfo::isNormal)
            .map(ScopeInfo::getDotName)
            .forEach(scope -> contextsForScope.merge(scope, 1, Integer::sum));
    // custom alterable contexts
    customAlterableContexts.getRegistered()
            .stream()
            .filter(cac -> Boolean.TRUE.equals(cac.isNormal))
            .map(cac -> cac.scopeAnnotation)
            .map(DotName::createSimple)
            .forEach(scope -> contextsForScope.merge(scope, 1, Integer::sum));

    return contextsForScope.entrySet()
            .stream()
            .filter(entry -> entry.getValue() == 1)
            .map(Map.Entry::getKey)
            .collect(Collectors.toSet());
}
```

Reply:
Sure, why not. Pls update the PR.

Reply:
Done, thanks!

Reply:
FTR, I also realized that custom alterable contexts are registered through `ContextRegistrar` as usual (`CustomAlterableContexts` only exists to generate `InjectableContext` implementation for custom `AlterableContext` implementations), so we don't need to care about them here. `BeanDeployment.customContexts` will include these contexts, so that's all we have to look at.